### PR TITLE
Java: support Java 22 language features

### PR DESCRIPTION
**Pull request by smowton**
New language features supported (all described by https://openjdk.org/jeps/456):

* Anonymous variables (e.g. `TypeName _`), which can occur in declarations and binding patterns.
* Anonymous lambda parameters
* Anonymous patterns (spelled simply `_`), which can occur in record-pattern context (e.g. `MyRecord(int x, _)`)
* Multiple patterns in one switch case statement, so long as no pattern declares any binding variables (`case A _, B _ when sharedGuard`)
* Fall-through to and from pattern-case statements that don't declare any binding variables (`case A _: case "Constant": case MyRecord(_):`)

Anonymous variables, lambda parameters and patterns only require pretty-printing and some formatting; they can naturally never take part in dataflow and their names are irrelevant in QL.

Most of the work goes into pattern cases gaining multiple possible patterns (which means the pattern will be entered if one or other pattern matches), and into implementing fall-through. This in particular has consequences for control flow, where pattern-case statements (which normally represent carrying out the type test(s)) need to be circumvented by fall-through flow (e.g. in the context `case A _: f(); case B _: g(); break;`, the successor of `f()` is `g()`, not a type test for `B`).

### CODE REVIEW

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -194,12 +194,14 @@ predicate simpleAstFlowStep(Expr e1, Expr e2) {
   // In the following three cases only record patterns need this flow edge, leading from the bound instanceof
   // or switch tested expression to a record pattern that will read its fields. Simple binding patterns are
   // handled via VariableAssign.getSource instead.
+  // We only consider unique patterns because cases that declare multiple patterns are not allowed to declare
+  // any identifiers, so can't participate in dataflow.
```
Comment (igfoo) [inline java/ql/lib/semmle/code/java/dataflow/internal/DataFlowUtil.qll:198]: For consistency, would it be better to only step into patterns that declare an identifier, so we don't step into `case X _:` but not `case X _, Y _:`?

File: java/ql/test/library-tests/pattern-switch/cfg/Test.java [hunk 2]
```diff
@@ -98,6 +98,33 @@ case A(B(var x, var y), var z):
...
+
+     switch ((String)thing) {
+       case "a":
+       case String _ when ((String)thing).length() == 5:
+       case "b":
+         break;
+       default:
+         break;
+     }
+
```
Comment (igfoo) [inline java/ql/test/library-tests/pattern-switch/cfg/Test.java:127]: As the end of a switch statement is special, I wonder if it is worth adding tests for something like
```
switch(thing) {
    case Integer:
    case String:
}
```
and
```
switch(thing) {
    case Float:
        break;
    case Integer:
    case String:
}

File: java/ql/lib/semmle/code/java/Statement.qll [hunk 3]
```diff
@@ -539,12 +539,29 @@ class ConstCase extends SwitchCase {
...
-  PatternCase() { pattern.isNthChildOf(this, 0) }
+  /**
+   * DEPRECATED: alias for getPatternAtIndex(0)
+   */
+  deprecated PatternExpr getPattern() { result = this.getPatternAtIndex(0) }
+
+  /**
+   * Gets this case's `n`th pattern.
+   */
+  PatternExpr getPatternAtIndex(int n) { result.isNthChildOf(this, n) }
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/lib/semmle/code/java/Statement.qll:552]: Can't overload the existing deprecated `getPattern`

[the remark this answers is not in the document]
Comment (igfoo) [reply java/ql/lib/semmle/code/java/Statement.qll:552]: Why not? It seems to work fine for me?

**Review by igfoo — APPROVED** [review]

**Review by igfoo — APPROVED** [review]

**Pull request merged**